mirrord/intproxy/src/background_tasks.rs

```diff
@@ -0,0 +1,193 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `task` - the [`BackgroundTask`] to be spawned and managed.
+    /// * `id` - unique identifier of the task.
+    /// * `channel_size` - size of [`mpsc`] channels used to communicate with the task.
+    ///
+    /// # Panics
+    ///
+    /// This method panics when attempting to register a task with a duplicate id.
+    pub fn register<T>(&mut self, task: T, id: Id, channel_size: usize) -> TaskSender<T>
+    where
+        T: 'static + BackgroundTask<MessageOut = MOut> + Send,
+        Err: From<T::Error>,
+        T::MessageIn: Send,
+    {
+        if self.streams.contains_key(&id) {
+            panic!("duplicate task id {id:?}");
+        }
+
+        let (in_msg_tx, in_msg_rx) = mpsc::channel(channel_size);
+        let (out_msg_tx, out_msg_rx) = mpsc::channel(channel_size);
+
+        self.streams.insert(
+            id.clone(),
+            StreamNotifyClose::new(ReceiverStream::new(out_msg_rx)),
+        );
+
+        let mut message_bus = MessageBus {
+            tx: out_msg_tx,
+            rx: in_msg_rx,
+        };
+
+        self.handles.insert(
+            id.clone(),
+            tokio::spawn(async move { task.run(&mut message_bus).await.map_err(Into::into) }),
+        );
+
+        TaskSender(in_msg_tx)
+    }
+
+    /// Returns the next update from one of registered tasks.
+    pub async fn next(&mut self) -> Option<(Id, TaskUpdate<MOut, Err>)> {
+        let (id, msg) = self.streams.next().await?;
+
+        let msg = match msg {
+            Some(msg) => (id, TaskUpdate::Message(msg)),
+            None => {
+                let res = self
+                    .handles
+                    .remove(&id)
+                    .expect("task handles and streams are out of sync")
+                    .await;
+                match res {
+                    Err(..) => (id, TaskUpdate::Finished(Err(TaskError::Panic))),
+                    Ok(res) => (id, TaskUpdate::Finished(res.map_err(TaskError::Error))),
+                }
+            }
+        };
+
+        Some(msg)
+    }
+
+    /// Waits for all registered tasks to finish and returns their results.
+    /// This method does not signalize the tasks to finish. Instead, one should drop all
+    /// [`TaskSender`]s first.
+    pub async fn results(self) -> Vec<(Id, Result<(), TaskError<Err>>)> {
+        std::mem::drop(self.streams);
+
+        let mut results = Vec::with_capacity(self.handles.len());
+        for (id, handle) in self.handles {
+            let result = match handle.await {
+                Err(..) => Err(TaskError::Panic),
+                Ok(res) => res.map_err(TaskError::Error),
+            };
+            results.push((id, result));
+        }
+
+        results
+    }
+}
+
+/// An error that can occur when executing a [`BackgroundTask`].
+#[derive(Debug)]
+pub enum TaskError<Err> {
+    /// An internal task error.
+    Error(Err),
+    /// A panic.
+    Panic,
```

Comment:
Not that it's a problem, but just so that I can understand - what is the benefit of separate variants for `Error` and `Panic`, and not represent panics as another type of error?

Reply:
Not sure if I understand. We need to return a single type for a task result. `BackgroundTask::run` method returns a `Result`, but the task can also panic.

Reply:
We could get rid of this extra enum if we add an extra requirement `BackgroundTask::Error: From<tokio::task::JoinError>`. Let me now what you think

Reply:
No, it's ok. Just whatever you think is better.